Intro R revamp (PR 2 of n) : First notebook (#166)
AlexsLemonade/training-modules: intro-to-R-tidyverse/01-introduction-base-R.Rmd

In other words, you can program in R without RStudio, but you can't use the RStudio environment without R.

For more information about RStudio than you ever wanted to know, see this [RStudio IDE Cheatsheet](https://d33wubrfki0l68.cloudfront.net/0dc0efccfaf638de8be4ce43e2bab8a91541d3d2/c4fdc/wp-content/uploads/2018/08/rstudio-ide.png).

## The RStudio Environment

The RStudio environment has four main **panes** (their specific location can be changed under Tools -> Global Options -> Pane Layout).
![RStudio Appearance](screenshots/rstudio-panes.png) 

1. The **editor** pane is where you can write R scripts and other documents. 
This is your _text editor_, which will allow you to save your R code for future use. 
Note that change code here will not run automatically until you runu it. 

2. The **console** pane is where you can _interactively_ run R code. 

3. The **environment** pane primarily displays the variables, sometimes known as _objects_ you define during a given R session. 

> 4. The **help viewer** pane has several tabs all of which are pretty important:

Author: `lower right` pane seems misleading because you can rearrange your panels and I've seen some participants do this.

> 4. The **help viewer** pane has several tabs all of which are pretty important:

Author: So while I understand that `help` viewer isn't a great name, I think just having something to call it helps even if it is not entirely encompassing of its functions.

> 4. The **help viewer** pane has several tabs all of which are pretty important:

Author: How about `Assistance tabs`?

> 4. The **help viewer** pane has several tabs all of which are pretty important:

Reviewer: "Assistance" might be fine. People do rearrange them, but this is the **default** arrangement, so maybe emphasize that... at this stage people should not have done much?

Future You comes into existence about one second after you write code, and has no idea what on earth Past You was thinking. 
Help out Future You by adding lots of comments! 
Future You next week thinks Today You is an idiot, and the only way you can convince Future You that Today You is reasonably competent is by adding comments in your code explaining why Today You is actually not so bad.

## Basic Calculations

### Mathematical operators

The most basic use of R is as a regular calculator:

| Operation | Symbol |
|-----------|--------|
| Add  | `+` | 
| Subtract  | `-` | 
| Multiply  | `*` | 
| Divide  | `/` | 
| Exponentiate | `^` or `**` | 

For example, we can do some simple multiplication like this. 
In this and following code chunks, the section with the gray background is R code, and the following white-background chunk is the code's output.

```{r}
5 * 6
```

### Defining and using variables 

To define a variable, we use the _assignment operator_ which looks like an arrow: `<-`, for example `x <- 7` takes the value on the right-hand side of the operator and assigns it to the variable name on the left-hand side. 

> The opposite is also possible, e.g. ` 7 -> x`, as long as the arrow is pointing towards the variable name.

Author: Okay agreed. This is a thing I've noticed Stephanie does and I don't necessarily like it but I thought maybe it would solidify the arrow concept of assignment. But I agree with you, don't want to encourage this.

```{r}
# Define a variable x to equal 7, and print out the value of x
x <- 7

> print(x)

Author: I don't either, Stephanie's original notebook was very pro `print` but I almost never use it, but wasn't sure how much we cared about this. Will switch to no `print`.

```

Some features of variables, considering the example `x <- 7`:
Every variable has a **name**, a **value**, and a **type**. 
This variable's name is `x`, its value is `7`, and its type is `numeric` (7 is a number!).
Re-defining a variable will overwrite the value.

```{r}
x <- 5.5

x
```

### Variable naming note:
As best you can, it is a good idea to make your variable names informative (e.g. `x` doesn't mean anything, but `cost_of_sandwich` is meaningful... if we're talking about sandwich prices, that is..). 

## Functions
We can use pre-built computation methods called "functions" for other operations. 

> In fact, we've already seen and used one - the `print()` function.

Author: Yup. Deleted this statement.

Functions have the following format, where the _argument_ is the information we are providing to the function for it to run.

```{r, eval=FALSE}
function_name(argument)
```

To learn about functions, we'll examine one called `log()` first. 

To know what a function does and how to use it, use the question mark which will reveal documentation in the **help pane**: `?log`
![rhelp](screenshots/rhelp-log.png) 

The documentation tells us that `log()` is derived from `{base}`, meaning it is a function that is part of base R. 
It provides a brief description of what the function does and shows several examples of to how use it.

In particular, documentation tells about how what argument(s) to provide:
+ The first _required_ argument is the value we'd like to take the log of, by default its _natural log_
+ The second _optional_ argument can specify a different base rather than the default `e`.

Functions also _return_ values for us to use. In the case of `log()`, the returned value is the log'd value the function computed.

> 

Author: I've added this but put them in their own chunks and added explanations. See what you think when I push this commit.

## Working with variables

> ### Variable Types

Author: Where do we want to put coercion then, a later notebook?

> ### Variable Types

Reviewer: My answer to many of these things is likely to be the same: "When we need it." If there is a reasonable case where we can motivate coercion in this notebook, that would make sense, but I feel like throwing it in without a real-world use case raises more questions than it solves.

> ### Variable Types

Author: Okay. So we'll table it for now and then figure it out later?

> ### Variable Types

Reviewer: That is my vote.

# Define a variable with a number
x <- 15
```

The function `class()` will tell us the variable's type.

```{r}
class(x)
```

Let's coerce it to a character. 

```{r}
x <- as.character(x)
class(x)
```

See it now has quotes around it? It's now a character and will behave as such

```{r}
x
```

Use this chunk to try to perform calculations with `x`, now that it is a character, what happens? 

```{r}
# Try to perform calculations on `x`
```

But we can't coerce everything:

```{r}
# Let's create a character variable
x <- "look at my character variable"
```

Let's try making this a numeric variable:

```{r, error=TRUE}
x <- as.numeric(x)
```

Print out `x`.

```{r}
x
```

R is telling us it doesn't know how to convert this to a numeric variable, so it has returned `NA` instead.

### Logical variables and operators

One of the most important variable types are _logical variables_ (known as "boolean" in many other languages). 
They help us to compare different quantities and there is a special set of operators for performing comparisons that _return logical values_:

| Operator | What it does |
|----------|--------------|
 `==`           |  Tests if two quantities are equal |

>   `>` and `<`   |  Tests if one quantity is greater than or less than another |

Reviewer: ```suggestion
  `>` and `<`   |  Tests if the left quantity is greater than or less than the one on the right |
```

  `>=` and `<=` |  Tests if one quantity is _greater than or equal to_ or _less than or equal to_ another |
`!`             |  Negates an operation |  

Let's examine their usage:

```{r, error=TRUE}
5 == 5

> 7 >= 9

Reviewer: ```suggestion
count <- 7
count >= 9
```

> 4 < 8

Reviewer: ```suggestion
4 < count
```

We can try out these functions on the vector `values_1_to_20` we've created. 

```{r}
mean(values_1_to_20)

# Try out some of the other functions we've listed above 
```

### A note on variable naming
We have learned functions such as `c`, `length`, `sum`, and etc. 
Imagine defining a variable called `c`: This will work, but it will lead to a 
lot of unintended bugs, so its best to avoid this. 

### The `%in%` logical operator 

`%in%` is useful for determing whether a given item(s) are in an vector.

```{r}
# is `7` in our vector? 
7 %in% values_1_to_20
```

```{r}
# is `50` in our vector? 
50 %in% values_1_to_20
```

We can test a vector of values being within another vector of values. 

```{r}
# Are these values in our vector? 
c(1:3, 7, 50) %in% values_1_to_20
```

## Data frames

_Data frames are the most fundamental unit of data analysis in R._ 
They are tables which consist of rows and columns, much like a _spreadsheet_. 
Each column is a variable which behaves as a _vector_, and each row is an observation. 
We will begin our exploration with the old trusted dataset `iris`, which comes with R. 
Learn about this dataset using the standard help approach of `?iris`.

> ### Exploring and indexing data frames

Reviewer: ```suggestion
### Exploring data frames
```

The first step to using any data is to look at it!!! 
RStudio contains a special function `View()` which allows you to literally view a variable.
You can click on the object in the environment pane. 

Some useful functions for exploring our data.frame include:

+ `head()` to see the first 6 rows of a data frame. Additional arguments supplied can change the number of rows.
+ `tail()` to see the last 6 rows of a data frame. Additional arguments supplied can change the number of rows.
+ `names()` to see the column names of the data frame.
+ `nrow()` to see how many rows are in the data frame
+ `ncol()` to see how many columns are in the data frame.

We can additionally explore _overall properties_ of the data frame with two different functions: `summary()` and `str()`.

This provides summary statistics for each column:

```{r}
summary(iris)
```

> This provides a short view of the contents of the data frame

Reviewer: ```suggestion
This provides a short view of the **str**ucture and contents of the data frame
```

Author: Nice.

```{r}
str(iris)
```

You'll notice that the column `Species` is a _factor_: This is a special type of character variable that represents distinct categories known as "levels". 
We have learned here that there are three levels in the `Species` column: setosa, versicolor, and virginica. 
We might want to explore individual columns of the data frame more in-depth. 

> We can _index_ these columns using the dollar sign `$`:

Reviewer: ```suggestion
We can examine individual columns using the dollar sign `$` to select one by name:
```

```{r}
# Extract Sepal.Length as a vector

> iris$Sepal.Length[1:10]

Reviewer: ```suggestion
iris$Sepal.Length

# indexing operators can be used too
iris$Sepal.Width[1:10]
```

```

We can perform our regular vector operations on columns directly.

```{r}
mean(iris$Sepal.Length)
```

> We can also achieve summary statistics for a single column directly. 

Reviewer: ```suggestion
We can also calculate the full summary statistics for a single column directly. 
```
